mvp dbc parsing

aaravm4 opened the pull request:
[Task](https://app.clickup.com/t/868czuawk)

Import cantools before running.

IshDeshpa commented on dbc/test_write/can0x01_recv_entries.h:
```diff
@@ -0,0 +1,159 @@
+CAN_RECV_ENTRY(0xc, 8, false)
```
can you put a comment next to each one with the actual message name

IshDeshpa commented on dbc/test_write/can0x01_recv_entries.h:
```diff
@@ -0,0 +1,159 @@
+CAN_RECV_ENTRY(0x25b, 1, false)
+CAN_RECV_ENTRY(0x25d, 8, false)
+CAN_RECV_ENTRY(0x29d, 4, false)
+CAN_RECV_ENTRY(0x2b4, 5, false)
```
also, the second number is supposed to be queue sizes, so they should all be the same and set to some default number that the user specifies. it's not the message length

IshDeshpa commented on dbc/test_write/can0x01_recv_entries.h:
prolly need to name the file just can1_recv_entries since that's what CAN pulls in specifically

IshDeshpa replied:
also yeah after running the script just output the new can1_recv_entries to either the directory they specify, or the same directory as the script, and then just tell the user on the CLI to move the header over to wherever they want

IshDeshpa commented on dbc/dbc.py:
```diff
@@ -0,0 +1,56 @@
+bold = "\033[1m"
+end = "\033[0m"
+
+if len(sys.argv) < 5: 
+    print(f"{bold} not enough args (dbc file path, can bus number, default size, output dir) {end}")
+    exit();
+
+DBC_FILE_PATH = Path(sys.argv[1])
+CAN_BUS = Path(sys.argv[2])
+DEFAULT_SIZE = Path(sys.argv[3])
+WRITE_FILE_DIR = Path(sys.argv[4])
```
output dir can be an optional argument with default being current directory. look at argparse library to make this easier

IshDeshpa commented on dbc/dbc.py:
```diff
@@ -0,0 +1,56 @@
+
+DBC_FILE_PATH = Path(sys.argv[1])
+CAN_BUS = Path(sys.argv[2])
+DEFAULT_SIZE = Path(sys.argv[3])
```
also yeah this default size will basically be what's set for all the messages' queue sizes since that's not gonna be specified in a DBC file and rather has to be specified by us.

IshDeshpa commented on dbc/dbc.py:
dbc-to-c.py perhaps? something that tells the user what it does

IshDeshpa reviewed (changes requested):
good job on the turnaround for this. a couple of big things but rest are mostly nitpicks.

Lakshay983 reviewed (changes requested):
Looks dope!

Also mayhaps add the cantools python stuff to a requirements.txt file and the docker can like build with it or smth. 

Can you make the utils.h header based on the bus instead of which peripheral. For example, the Controls peripheral canbus should be a different file IMO. It is technically separated by the can peripheral but I think it's better for organization to be separated by bus. There's a bus parameter in the DBC format you can use.

Can you also error handle on repeat IDs on the same DBC. idk if cantools does that automatically but it would be nice to display if there's an error since we could easily mess it up

Last thing, in your dbc.py file can you add an argument to only verify that it's a valid CAN DBC with no repeat IDs. so like i'd run python3 dbc.py --verify and it won't generate the dbc, it'll just say if it's a valid DBC or not. That way we can set that up with a github action to verify DBC stuff on push

LGTM other than what Ishan posted

Two things that may be for a different pr since ig this is mostly for an mvp
- It'd be nice to have like a macro bit mask based on the signal. So like if a signal is the first 4 bits of a CAN message you can just & the message with the bit mask to get the signal
- There's a value table field for signals that tell you what the signal's values mean. ie a 1 means safe or a 0 means unsafe. It'd be cool to have that as a bunch of enums.

aaravm4 commented:
@IshDeshpa @Lakshay983 addressed PR comments.

(1) Cantools checks for duplicates everytime you load a `.dbc` file. It displays "`Overwriting X with X`" on the terminal if duplicates occur so there's isn't really a need for  `--verify` command.

(2) Made `#define` function macros to have bitwise `&` and bit shifting. Thought it would make life easier.

(3) Created basic `enum` declarations without defining a type. Lmk if that's needed.

👍

IshDeshpa reviewed (changes requested):
Just some small clerical stuff but really good work here!

Lakshay983 reviewed (approved):
LGTM other than these changes + Ishan's changes. I also think that you should make the -p flag the default option

Also make sure to fix the merge conflict with Dockerfile

You're going to remove all the file_examples right? 

Before this gets merged you should testrun this with BSP_CAN to make sure all your enum + define + recv_entry stuff works well

IshDeshpa reviewed (dismissed):
Jolly good show!

Lakshay983 reviewed (dismissed):
LGTM once this gets tested with BSP CAN

IshDeshpa commented:
make sure to merge in main here

IshDeshpa commented:
If we have a can directory on main now, can we make the dbc directory a subdirectory of that?

Lakshay983 commented:
> If we have a can directory on main now, can we make the dbc directory a subdirectory of that?

I say just rename the main directory to be DBC. We're gonna delete the sheets soon anyway
